### ISSUE TITLE
container can accept requests when became unhealthy in swarm mode

### ISSUE THREAD

**Issue by ziqingzhu** [docker/docker#28741]
**Description**

**Steps to reproduce the issue:**
1.create a service in swarm mode
2.post request to this service continue
3.update service 
4.one of container in this service health became (unhealthy), after a while,it will shutdown and another container will be create

**Describe the results you received:**
when the container became unhealthy , it can also accept request

**Describe the results you expected:**
when the container became unhealthy, it should be can not accept request

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
Client:
 Version:      1.12.3
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   6b644ec
 Built:        
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.3
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   6b644ec
 Built:        
 OS/Arch:      linux/amd64

```

**Output of `docker info`:**

```
Containers: 13
 Running: 6
 Paused: 0
 Stopped: 7
Images: 5
Server Version: 1.12.3
Storage Driver: devicemapper
 Pool Name: docker-8:3-6815748-pool
 Pool Blocksize: 65.54 kB
 Base Device Size: 10.74 GB
 Backing Filesystem: xfs
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 2.374 GB
 Data Space Total: 107.4 GB
 Data Space Available: 105 GB
 Metadata Space Used: 5.026 MB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.142 GB
 Thin Pool Minimum Free Space: 10.74 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Data loop file: /data/docker/devicemapper/devicemapper/data
 WARNING: Usage of loopback devices is strongly discouraged for production use. Use `--storage-opt dm.thinpooldev` to specify a custom block storage device.
 Metadata loop file: /data/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.107-RHEL7 (2016-06-09)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: null bridge host overlay
Swarm: active
 NodeID: dxx0lvd058ykf2a3rix9xqs3x
 Is Manager: true
 ClusterID: 5394pay22kohc4suti8481ngk
 Managers: 2
 Nodes: 2
 Orchestration:
  Task History Retention Limit: 10
 Raft:
  Snapshot Interval: 5000
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
 Node Address: 10.12.8.44
Runtimes: runc
Default Runtime: runc
Security Options: seccomp
Kernel Version: 4.8.10-1.el7.elrepo.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 56
Total Memory: 125.7 GiB
Name: s2-core-044
ID: WSCQ:2DTP:M4PC:VVND:ISVW:OTD4:QM2S:OSNR:XCID:3HLR:DZMA:DK7L
Docker Root Dir: /data/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 86
 Goroutines: 245
 System Time: 2016-11-23T17:03:58.539044083+08:00
 EventsListeners: 6
Registry: https://index.docker.io/v1/
WARNING: bridge-nf-call-ip6tables is disabled
Insecure Registries:
 docker.v2.aispeech.com
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
physical 
Centos7, 4.8.10-1.el7.elrepo.x86_64

**Comment by rieske** [comment]
I am experiencing the same.
Docker `1.12.3`
CentOS `3.10.0-327.22.2.el7.x86_64`

I have healthchecks set up in Dockerfiles and verified that they are working correctly.
When I do a rolling update of a service in the swarm cluster however, routing mesh routes requests to a replica which is not healthy yet and still in `starting` state.

My expectation is that it should route requests only to healthy services (Running state)
This is a blocker for rolling zero-downtime upgrades.

**Comment by rieske** [comment]
This is claimed to be fixed in `1.12.3` https://github.com/docker/docker/issues/26067
But to me this is not the case.

**Comment by thaJeztah** [comment]
This looks like a duplicate of https://github.com/docker/docker/issues/24092, which is implemented in https://github.com/docker/docker/pull/27279 for docker 1.13. However, that PR does not explicitly mention _removing_ containers from the load balancer if health checks fail. 

@ziqingzhu @rieske would you be able to test the 1.13 release candidate to see if that addresses your issue? Download / install instructions can be found on the releases page; https://github.com/docker/docker/releases. Keep in mind that these are release candidates, and may still contain bugs, so should not be installed for "production", or on machines that contain critical data :smile:

**Comment by thaJeztah** [comment]
Also /cc @dongluochen PTAL

**Comment by rieske** [comment]
Seems to be working nicely on `Docker version 1.13.0-rc2, build 1f9b3ef`

**Comment by thaJeztah** [comment]
Thanks for testing @rieske!

@ziqingzhu I'll go ahead and close this issue because it should be addressed in 1.13, but let me know if you can still reproduce with that version, and I'll reopen.

**Comment by ziqingzhu** [comment]
Great! It is working as expected~
